File: examples/schema/json-schema/fields.ts

// tslint:disable: object-literal-key-quotes
const schema = {
  "$schema": "http://json-schema.org/draft-06/schema#",
  "definitions": {
    "NumberType": {
      "type": "number"
    },
    "A": {
      "type": "object",
      "properties": {
        "numberTypeField": {
          "$ref": "#/definitions/NumberType"
        },
        "integerTypeField": {
          "type": "integer"
        },
        "bigIntTypeField": {
          "type": [
            "integer",
            "string"
          ]
        },
        "stringTypeField": {
          "type": "string"
        },
        "booleanTypeField": {
          "type": "boolean"
        },
        "nullTypeField": {
          "type": "null"
        },
        "undefinedTypeField": {
          "type": "null"
        },
        "anyTypeField": {
          "type": [
            "null",
            "integer",
            "number",
            "string",
            "boolean",
            "array",
            "object"
          ]
        },
        "unknownTypeField": {
          "type": [
            "null",
            "integer",
            "number",
            "string",
            "boolean",
            "array",
            "object"
          ]
        },
        "neverTypeField": {
          "type": "null"
        },
        "numberValueTypeField": {
          "type": "number",
          "enum": [
            3
          ]
        },
        "integerValueTypeField": {
          "type": "integer"
        },
        "bigIntValueTypeField": {
          "type": [
            "integer",
            "string"
          ],
          "enum": [
            "7",
            7
          ]
        },
        "stringValueTypeField": {
          "type": "string",
          "enum": [
            "XB"
          ]
        },
        "booleanValueTypeField": {
          "type": "boolean",
          "enum": [
            true
          ]
        }
      },
      "required": [
        "numberTypeField",
        "integerTypeField",
        "bigIntTypeField",
        "stringTypeField",
        "booleanTypeField",
        "nullTypeField",
        "undefinedTypeField",
        "anyTypeField",
        "unknownTypeField",
        "neverTypeField",
        "numberValueTypeField",
        "integerValueTypeField",
        "bigIntValueTypeField",
        "stringValueTypeField",
        "booleanValueTypeField"
      ],
      "additionalProperties": false
    },
    "B": {
      "type": "object",
      "properties": {
        "numberTypeField": {
          "$ref": "#/definitions/NumberType"
        },
        "integerTypeField": {
          "type": "integer"
        },
        "bigIntTypeField": {
          "type": [
            "integer",
            "string"
          ]
        },
        "stringTypeField": {
          "type": "string"
        },
        "booleanTypeField": {
          "type": "boolean"
        },
        "nullTypeField": {
          "type": "null"
        },
        "undefinedTypeField": {
          "type": "null"
        },
        "anyTypeField": {
          "type": [
            "null",
            "integer",
            "number",
            "string",
            "boolean",
            "array",
            "object"
          ]
        },
        "unknownTypeField": {
          "type": [
            "null",
            "integer",
            "number",
            "string",
            "boolean",
            "array",
            "object"
          ]
        },
        "neverTypeField": {
          "type": "null"
        },
        "numberValueTypeField": {
          "type": "number",
          "enum": [
            3
          ]
        },
        "integerValueTypeField": {
          "type": "integer"
        },
        "bigIntValueTypeField": {
          "type": [
            "integer",
            "string"
          ],
          "enum": [
            "7",
            7
          ]
        },
        "stringValueTypeField": {
          "type": "string",
          "enum": [
            "XB"
          ]
        },
        "booleanValueTypeField": {
          "type": "boolean",
          "enum": [
            true
          ]
        }
      },
      "additionalProperties": false
    },
    "C": {
      "type": "object",
      "properties": {
        "numberTypeField": {
          "$ref": "#/definitions/NumberType"
        },
        "integerTypeField": {
          "type": "integer"
        },
        "bigIntTypeField": {
          "type": [
            "integer",
            "string"
          ]
        },
        "stringTypeField": {
          "type": "string"
        },
        "booleanTypeField": {
          "type": "boolean"
        },
        "nullTypeField": {
          "type": "null"
        },
        "undefinedTypeField": {
          "type": "null"
        },
        "anyTypeField": {
          "type": [
            "null",
            "integer",
            "number",
            "string",
            "boolean",
            "array",
            "object"
          ]
        },
        "unknownTypeField": {
          "type": [
            "null",
            "integer",
            "number",
            "string",
            "boolean",
            "array",
            "object"
          ]
        },
        "neverTypeField": {
          "type": "null"
        },
        "numberValueTypeField": {
          "type": "number",
          "enum": [
            3
          ]
        },
        "integerValueTypeField": {
          "type": "integer"
        },
        "bigIntValueTypeField": {
          "type": [
            "integer",
            "string"
          ],
          "enum": [
            "7",
            7
          ]
        },
        "stringValueTypeField": {
          "type": "string",
          "enum": [
            "XB"
          ]
        },
        "booleanValueTypeField": {
          "type": "boolean",
          "enum": [
            true
          ]
        }
      },
      "required": [
        "numberTypeField",
        "integerTypeField",
        "bigIntTypeField",
        "stringTypeField",
        "booleanTypeField",
        "nullTypeField",
        "undefinedTypeField",
        "anyTypeField",
        "unknownTypeField",
        "neverTypeField",
        "numberValueTypeField",
        "integerValueTypeField",
        "bigIntValueTypeField",
        "stringValueTypeField",
        "booleanValueTypeField"
      ],
      "additionalProperties": false
    },
    "D": {
      "type": "object",
      "properties": {
        "numberTypeField": {
          "$ref": "#/definitions/NumberType"
        },
        "integerTypeField": {
          "type": "integer"
        },
        "bigIntTypeField": {
          "type": [
            "integer",
            "string"
          ]
        },
        "stringTypeField": {
          "type": "string"
        },
        "booleanTypeField": {
          "type": "boolean"
        },
        "nullTypeField": {
          "type": "null"
        },
        "undefinedTypeField": {
          "type": "null"
        },
        "anyTypeField": {
          "type": [
            "null",
            "integer",
            "number",
            "string",
            "boolean",
            "array",
            "object"
          ]
        },
        "unknownTypeField": {
          "type": [
            "null",
            "integer",
            "number",
            "string",
            "boolean",
            "array",
            "object"
          ]
        },
        "neverTypeField": {
          "type": "null"
        },
        "numberValueTypeField": {
          "type": "number",
          "enum": [
            3
          ]
        },
        "integerValueTypeField": {
          "type": "integer"
        },
        "bigIntValueTypeField": {
          "type": [
            "integer",
            "string"
          ],
          "enum": [
            "7",
            7
          ]
        },
        "stringValueTypeField": {
          "type": "string",
          "enum": [
            "XB"
          ]
        },
        "booleanValueTypeField": {
          "type": "boolean",
          "enum": [
            true
          ]
        }
      },
      "additionalProperties": false
    },
    "E": {
      "type": "object",
      "properties": {},
      "patternProperties": {
        "^.*$": {
          "type": [
            "null",
            "integer",
            "number",
            "string",
            "boolean",
            "array",
            "object"
          ],
          "description": "additional props"
        }
      },
      "additionalProperties": false
    },
    "Z1": {
      "type": "object",
      "properties": {
        "foo": {
          "$ref": "#/definitions/ACL/properties/target"
        },
        "bar": {
          "type": "string",
          "minLength": 0
        },
        "baz": {
          "$ref": "#/definitions/ACL/properties/target"
        }
      },
      "required": [
        "foo",
        "bar",
        "baz"
      ],
      "additionalProperties": false
    },
    "ACL": {
      "type": "object",
      "properties": {
        "target": {
          "type": "string"
        },
        "value": {
          "type": "string",
          "minLength": 0
        }
      },
      "required": [
        "target",
        "value"
      ],
      "additionalProperties": false
    },
    "Z2": {
      "type": "object",
      "properties": {
        "foo": {
          "$ref": "#/definitions/ACL/properties/target"
        },
        "bar": {
          "type": "string",
          "minLength": 0
        },
        "baz": {
          "type": "string",
          "maxLength": 10
        }
      },
      "required": [
        "foo",
        "bar",
        "baz"
      ],
      "additionalProperties": false
    },
    "ErrorTypes": {
      "type": [
        "string",
        "number"
      ],
      "enum": [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9
      ]
    },
    "Foo": {
      "type": "object",
      "properties": {
        "name": {
          "type": "string",
          "pattern": "^[A-Za-z]+$"
        },
        "email": {
          "type": "string",
          "pattern": "^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\\.[a-zA-Z0-9-]+)*$"
        }
      },
      "required": [
        "name",
        "email"
      ],
      "additionalProperties": false
    },
    "Bar": {
      "type": "object",
      "properties": {
        "foo": {
          "$ref": "#/definitions/Foo"
        }
      },
      "required": [
        "foo"
      ],
      "additionalProperties": false
    },
    "Baz": {
      "type": "object",
      "properties": {
        "aaa1": {
          "type": "string"
        },
        "aaa2": {
          "type": "string"
        },
        "aaa3": {
          "type": "string"
        },
        "aaa4": {
          "type": "string"
        },
        "bbb1": {
          "type": "integer",
          "minimum": -3,
          "maximum": 5
        },
        "bbb2": {
          "type": "integer",
          "minimum": -3,
          "maximum": 5
        },
        "bbb3": {
          "type": "integer",
          "maximum": -3
        },
        "bbb4": {
          "type": "integer",
          "minimum": 5
        },
        "ccc1": {
          "type": "number",
          "minimum": -3,
          "maximum": 5,
          "description": "comment"
        },
        "ccc2": {
          "type": "number",
          "minimum": -3,
          "maximum": 5,
          "description": "comment"
        },
        "ccc3": {
          "type": "number",
          "maximum": -3,
          "description": "comment"
        },
        "ccc4": {
          "type": "number",
          "minimum": 5,
          "description": "comment"
        },
        "ddd1": {
          "type": "array",
          "items": {
            "anyOf": [
              {
                "type": "number"
              },
              {
                "type": "string"
              }
            ]
          }
        },
        "ddd2": {
          "type": "array",
          "items": {
            "anyOf": [
              {
                "type": "number"
              },
              {
                "type": "string"
              }
            ]
          }
        },
        "ddd3": {
          "type": "array",
          "items": {
            "anyOf": [
              {
                "type": "number"
              },
              {
                "type": "string"
              }
            ]
          }
        },
        "ddd4": {
          "type": "array",
          "items": {
            "anyOf": [
              {
                "type": "number"
              },
              {
                "type": "string"
              }
            ]
          }
        },
        "ddd5": {
          "type": "array",
          "items": {
            "anyOf": [
              {
                "type": "number"
              },
              {
                "type": "string"
              }
            ]
          }
        },
        "ddd6": {
          "type": "array",
          "items": {
            "anyOf": [
              {
                "oneOf": [
                  {
                    "type": "number"
                  },
                  {
                    "type": "null"
                  }
                ]
              },
              {
                "type": "string"
              }
            ]
          }
        },
        "eee1": {
          "type": "array",
          "items": {
            "type": "string"
          }
        },
        "eee2": {
          "type": "array",
          "items": {
            "type": "string"
          },
          "minItems": 10
        },
        "eee3": {
          "type": "array",
          "items": {
            "type": "string"
          },
          "maxItems": 20
        },
        "eee4": {
          "type": "array",
          "items": {
            "type": "string"
          },
          "minItems": 10,
          "maxItems": 20
        },
        "fff": {
          "$ref": "#/definitions/ErrorTypes"
        }
      },
      "required": [
        "aaa1",
        "aaa2",
        "aaa3",
        "aaa4",
        "bbb1",
        "bbb2",
        "bbb3",
        "bbb4",
        "ccc1",
        "ccc2",
        "ccc3",
        "ccc4",
        "ddd1",
        "ddd2",
        "ddd3",
        "ddd4",
        "ddd5",
        "ddd6",
        "eee1",
        "eee2",
        "eee3",
        "eee4",
        "fff"
      ],
      "additionalProperties": false
    },
    "User": {
      "type": "object",
      "properties": {
        "userName": {
          "type": "string",
          "pattern": "^[A-Za-z]+$"
        },
        "primaryEmail": {
          "type": "string",
          "pattern": "^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\\.[a-zA-Z0-9-]+)*$"
        },
        "primaryAliasName": {
          "type": "string",
          "pattern": "^[A-Za-z]+$"
        },
        "aliasNames": {
          "type": "array",
          "items": {
            "type": "string",
            "pattern": "^[A-Za-z]+$"
          }
        }
      },
      "required": [
        "userName",
        "primaryEmail",
        "primaryAliasName",
        "aliasNames"
      ],
      "additionalProperties": false
    }
  }
};
export default schema;
// tslint:enable: object-literal-key-quotes

File: examples/schema/csharp/fields.cs
using System.ComponentModel.DataAnnotations;

namespace Tynder.UserSchema
{
    using NumberType = System.Object;


    public class A 
    {
        [Required]
        public NumberType numberTypeField { get; set; }

        public int integerTypeField { get; set; }

        public decimal bigIntTypeField { get; set; }

        [Required]
        public string stringTypeField { get; set; }

        public bool booleanTypeField { get; set; }

        [Required]
        public object nullTypeField { get; set; }

        [Required]
        public object undefinedTypeField { get; set; }

        [Required]
        public object anyTypeField { get; set; }

        [Required]
        public object unknownTypeField { get; set; }

        [Required]
        public object neverTypeField { get; set; }

        public double numberValueTypeField { get; set; }

        public int integerValueTypeField { get; set; }

        [Required]
        public object bigIntValueTypeField { get; set; }

        [Required]
        public string stringValueTypeField { get; set; }

        public bool booleanValueTypeField { get; set; }
    }


    public class B 
    {
        public NumberType numberTypeField { get; set; }

        public int? integerTypeField { get; set; }

        public decimal? bigIntTypeField { get; set; }

        public string stringTypeField { get; set; }

        public bool? booleanTypeField { get; set; }

        public object nullTypeField { get; set; }

        public object undefinedTypeField { get; set; }

        public object anyTypeField { get; set; }

        public object unknownTypeField { get; set; }

        public object neverTypeField { get; set; }

        public double? numberValueTypeField { get; set; }

        public int? integerValueTypeField { get; set; }

        public object bigIntValueTypeField { get; set; }

        public string stringValueTypeField { get; set; }

        public bool? booleanValueTypeField { get; set; }
    }


    public class C : A 
    {
    }


    public class D 
    {
        public NumberType numberTypeField { get; set; }

        public int? integerTypeField { get; set; }

        public decimal? bigIntTypeField { get; set; }

        public string stringTypeField { get; set; }

        public bool? booleanTypeField { get; set; }

        public object nullTypeField { get; set; }

        public object undefinedTypeField { get; set; }

        public object anyTypeField { get; set; }

        public object unknownTypeField { get; set; }

        public object neverTypeField { get; set; }

        public double? numberValueTypeField { get; set; }

        public int? integerValueTypeField { get; set; }

        public object bigIntValueTypeField { get; set; }

        public string stringValueTypeField { get; set; }

        public bool? booleanValueTypeField { get; set; }
    }


    public class E 
    {
    }


    public class Z1 
    {
        [Required]
        public string foo { get; set; }

        [Required, MinLength(0)]
        public string bar { get; set; }

        [Required]
        public string baz { get; set; }
    }


    public class ACL 
    {
        [Required]
        public string target { get; set; }

        [Required, MinLength(0)]
        public string value { get; set; }
    }


    public class Z2 
    {
        [Required]
        public string foo { get; set; }

        [Required, MinLength(0)]
        public string bar { get; set; }

        [Required, MaxLength(10)]
        public string baz { get; set; }
    }


    public static class ErrorTypes
    {
        /** comment */
        public static double InvalidDefinition { get { return 1; } }

        /** comment */
        public static double Required { get { return 2; } }

        /** comment */
        public static double TypeUnmatched { get { return 3; } }

        /** comment */
        public static double RepeatQtyUnmatched { get { return 4; } }

        public static double SequenceUnmatched { get { return 5; } }

        public static double ValueRangeUnmatched { get { return 6; } }

        public static double ValuePatternUnmatched { get { return 7; } }

        public static double ValueLengthUnmatched { get { return 8; } }

        public static double ValueUnmatched { get { return 9; } }

        public static double Aaaaa { get { return 99; } }

        public static string Bbbbb { get { return "string bbbbb"; } }
    }


    public static class ErrorTypes2
    {
        public static double Aaaaa { get { return 99; } }
    }


    public static class ErrorTypes3
    {
        public static double Zzzzz { get { return 0; } }

        public static double Aaaaa { get { return 99; } }
    }


    public class Foo 
    {
        [Required, RegularExpression(@"^[A-Za-z]+$")]
        public string name { get; set; }

        [Required, RegularExpression(@"^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$")]
        public string email { get; set; }
    }


    public class Bar 
    {
        [Required]
        public Foo foo { get; set; }
    }


    public class Baz 
    {
        [Required, Range(typeof(string), "a", "z")]
        public string aaa1 { get; set; }

        [Required, Range(typeof(string), "a", "z")]
        public string aaa2 { get; set; }

        [Required, Range(typeof(string), "", "z")]
        public string aaa3 { get; set; }

        [Required, Range(typeof(string), "a", "\U00010FFFF")]
        public string aaa4 { get; set; }

        [Range((int)-3, (int)5)]
        public int bbb1 { get; set; }

        [Range((int)-3, (int)5)]
        public int bbb2 { get; set; }

        [Range(Int32.MinValue, (int)-3)]
        public int bbb3 { get; set; }

        [Range((int)5, Int32.MaxValue)]
        public int bbb4 { get; set; }

        /** comment */
        [Range((double)-3, (double)5)]
        public double ccc1 { get; set; }

        /** comment */
        [Range((double)-3, (double)5)]
        public double ccc2 { get; set; }

        /** comment */
        [Range(Double.MinValue, (double)-3)]
        public double ccc3 { get; set; }

        /** comment */
        [Range((double)5, Double.MaxValue)]
        public double ccc4 { get; set; }

        [Required]
        public object[] ddd1 { get; set; }

        [Required]
        public object[] ddd2 { get; set; }

        [Required]
        public object[] ddd3 { get; set; }

        [Required]
        public object[] ddd4 { get; set; }

        [Required]
        public object[] ddd5 { get; set; }

        [Required]
        public object[] ddd6 { get; set; }

        [Required]
        public string[] eee1 { get; set; }

        [Required, MinLength(10)]
        public string[] eee2 { get; set; }

        [Required, MaxLength(20)]
        public string[] eee3 { get; set; }

        [Required, MinLength(10), MaxLength(20)]
        public string[] eee4 { get; set; }

        [Required]
        public object fff { get; set; }

        [Required]
        public Bar ggg1 { get; set; }

        [Required]
        public Bar[] ggg2 { get; set; }

        [Required]
        public Bar[][] ggg3 { get; set; }
    }


    public class User 
    {
        [Required, RegularExpression(@"^[A-Za-z]+$")]
        public string userName { get; set; }

        [Required, RegularExpression(@"^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$")]
        public string primaryEmail { get; set; }

        [Required, RegularExpression(@"^[A-Za-z]+$")]
        public string primaryAliasName { get; set; }

        [Required]
        public string[] aliasNames { get; set; }
    }
}
